### PR TITLE
[sdk#713] Add NS, NSE connect clients

### PR DESCRIPTION
**Pull request by Bolodya1997**
## Description
Adds NS, NSE connect clients.

## Issue link
#713 

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [x] Refactoring
- [ ] CI

### CODE REVIEW

**Comment by edwarnicke** [comment]
Why are we creating per use case version of connect clients?

**Comment by Bolodya1997** [comment]
> Why are we creating per use case version of connect clients?

It is not a special per use case version of connect client, it is a first step of reworking to new connect approach :)
Actually connect server over adapter is not applicable for registry healing cases because it is a source of a lot of unstable behavior related to how we handle client -> server and server -> client gRPC stream translations in adapters. So we can make the following steps:
1. Create a connect client for NSC chain and make registry healing working with it.
2. Rework connect server to work with this connect client, remove an old one, merge `connect` and `connectto` packages.

**Comment by denis-tingaikin** [comment]
Can we go with step 2 immediately?

File: pkg/tools/grpcutils/url.go [hunk 1]
```diff
@@ -93,3 +93,15 @@ func TargetToNetAddr(target string) (network, addr string) {
 
 	return network, target
 }
+
+// CheckURLFree returns check for is given url is free for Listen
```
Comment (denis-tingaikin) [inline pkg/tools/grpcutils/url.go:97]: I don't think that this support any url, as I can see only tcp

**Review by denis-tingaikin — APPROVED** [review]
In general looks good

File: pkg/registry/common/connect/ns_client.go [hunk 2]
```diff
@@ -20,85 +20,121 @@ import (
...
+	lock sync.RWMutex
 }
 
-func (c *connectNSClient) init() error {
-	c.initOnce.Do(func() {
-		ctx, cancel := context.WithCancel(c.ctx)
-		c.ctx = ctx
+// NewNetworkServiceRegistryClient returns a new NS registry client chain element connecting to the remote
+//                                 NS registry server
+func NewNetworkServiceRegistryClient(ctx context.Context, connectTo string, opts ...Option) registry.NetworkServiceRegistryClient {
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply pkg/registry/common/connect/ns_client.go]: Yes, it looks better. Done.

File: pkg/tools/grpcutils/url.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/tools/grpcutils/url.go]: Fixed to support any URL.

File: pkg/tools/grpcutils/url.go [hunk 3]
```diff
@@ -93,3 +93,15 @@ func TargetToNetAddr(target string) (network, addr string) {
...
+func CheckURLFree(u string) func() bool {
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply pkg/tools/grpcutils/url.go]: I have slightly refactored this method:
```go
func CheckURLFree(u *url.URL) bool {
	ln, err := net.Listen(TargetToNetAddr(URLToTarget(u)))
	if err == nil {
		err = ln.Close()
	}
	return err == nil
}
```
So probably in such way it is OK to locate it in `grpcutils`?

Actually I can't see a better place for storing some utility method for gRPC action even if this action is mostly consumed in tests.

**Comment by edwarnicke** [comment]
> > Why are we creating per use case version of connect clients?
> 
> It is not a special per use case version of connect client, it is a first step of reworking to the new connect approach :)
> Actually connect server over adapter is not applicable for registry healing cases because it is a source of a lot of unstable behavior related to how we handle client -> server and server -> client gRPC stream translations in adapters. So we can make the following steps:
> 
> 1. Create a connect client for NSC chain and make registry healing working with it.
> 2. Rework connect server to work with this connect client, remove an old one, merge `connect` and `connectto` packages.

Apologies, I read this too fast, and misread on a quick skim through the code, and failed to note this was for registry connect.
I had thought we were expanding our connection of connects for networkservice.  Taking the time to read more closely, this is fine :)

**Pull request merged**